plugins/BEdita/Core/src/Model/Action/SignupUserAction.php

```diff
@@ -173,6 +184,45 @@ protected function createUser(array $data)
         ]);
     }
 
+    /**
+     * Add roles to user if requested, with validity check
+     *
+     * @param \BEdita\Core\Model\Entity\User $entity The user created
+     * @param array $data The signup data
+     * @return void
+     */
+    protected function addRoles(User $entity, array $data)
+    {
+        if (empty($data['roles'])) {
+            return;
+        }
+        $roles = $this->loadRoles($data['roles']);
+        $association = $this->Users->associations()->getByProperty('roles');
+        $association->link($entity, $roles);
+    }
+
+    /**
+     * Load requested roles entities with validation
+     *
+     * @param array $roles Requested role names
+     * @return \BEdita\Core\Model\Entity\Role[] requested role entities
+     * @throws \Cake\Network\Exception\BadRequestException When role validation fails
+     */
+    protected function loadRoles(array $roles)
+    {
+        $entities = [];
+        $allowed = (array)Configure::read('Signup.roles');
```

Comment:
I think a better place to check roles should be an application rule. Since `SignupUserAction` implements `EventListenerInterface` it should be easy as

```
/**
 * {@inheritdoc}
 */
public function implementedEvents()
{
    return [
        'Auth.signup' => 'sendMail',
        'Model.buildRules' => 'isRolesValid',
    ];
}
```

and implement `isRolesValid` method.

Reply:
we'll do that after merge, if not a problem